In [4]:
import pandas as pd
import numpy as np
from scipy.stats import bartlett
df = pd.read_csv("dataset/diamonds.csv")
stat, p = bartlett(df.loc[df["cut"] == "Good", "price"],
                  df.loc[df["cut"] == "Ideal", "price"],
                  df.loc[df["cut"] == "Premium", "price"])

In [5]:
stat #검정통계량

368.0903292307125

In [6]:
p #p-value

1.175440327456315e-80

세공수준(Cut)별 가격(Price)의 분산은 등분산 검정결과 p-value가 $1.1754 \times 10^ {-80}$ 으로 유의수준 5%(0.05)기준으로 유의수준보다 p-value가 작기 때문에 귀무가설을 기각하고 대립가설을 채택함. 즉 집단간 적어도 한 쌍의 분산은 서로 다름.

### 단일표본T검정

In [7]:
from scipy.stats import ttest_1samp
df = pd.read_csv("dataset/iris_xlsx.csv")

In [9]:
stat, p = ttest_1samp(df["Sepal.Length"], popmean = 4)
print(round(stat, 3))
print(round(p, 3))

27.264
0.0


p-value가 0으로 유의수준 5%(00.5) 기준 p-value가 유의수준보다 작기 때문에 귀무가설을 기각하고 대립가설을 채택하여 모평균 4와 붓꽃(iris)의 꽃받침 길이(Sepal.Length)의 평균은 서로 유의미하게 차이가 난다고 할 수있다.

In [10]:
round(df["Sepal.Length"].mean(), 3)

5.843

In [35]:
stat, p = ttest_1samp(df["Sepal.Length"], popmean = 5.6)
print(round(stat, 3))
print(round(p, 5))

3.599
0.00043


### 대응표본T검정

In [42]:
from scipy.stats import ttest_rel
stat, p = ttest_rel(df["Sepal.Length"], df["Sepal.Width"])
print(stat, p)

34.81519257082818 1.8495537347987982e-73


p-value가 0으로 유의수준 5%(00.5) 기준 p-value가 유의수준보다 작기 때문에 귀무가설을 기각하고 대립가설을 채택하여 평균은 서로 유의미하게 차이가 난다고 할 수있다.

### 독립2표본T검정

In [44]:
from scipy.stats import ttest_ind
stat, p = ttest_ind(df["Sepal.Length"], df["Sepal.Width"], equal_var=True)
print(stat, p)

36.463283934449116 7.027919392009369e-112


p-value가 0으로 유의수준 5%(00.5) 기준 p-value가 유의수준보다 작기 때문에 귀무가설을 기각하고 대립가설을 채택하여 평균은 서로 유의미하게 차이가 난다고 할 수있다.

### 독립성 검정
* 입력은 두 개의 명목형 변수의 각 원소의 빈도
* 출력은 검정통계량, P-value, 자유도, 기대도수 => Tuple

In [47]:
from scipy.stats import chi2_contingency
df = pd.read_csv("dataset/traffic_accident.txt", sep="\t")
df

seq sex serious_yn  age
0        1   M          N    3
1        2   M          N    6
2        3   M          N    4
3        4   M          N    4
4        5   F          N    3
...    ...  ..        ...  ...
1062  1063   F          N    6
1063  1064   M          N    4
1064  1065   M          N    6
1065  1066   F          N    6
1066  1067   F          N    3

[1067 rows x 4 columns]

In [48]:
chi2_contingency(pd.crosstab(df["serious_yn"], df["sex"]), correction=False) # crosstab 주의

(4.248074934656112,
 0.03929484927552679,
 1,
 array([[286.99343955, 679.00656045],
        [ 30.00656045,  70.99343955]]))

In [49]:
pd.crosstab(df["serious_yn"], df["sex"])

sex           F    M
serious_yn          
N           296  670
Y            21   80